### PR TITLE
Async Indexer For Stock

### PR DESCRIPTION
**Pull request by nuzil**
### Description (*)
Cause Stock / Source reindex is Synchronous operation, it lead to Deadlocks when multiple operations managing content.

### Follow UP

Colin Macdonald  11 days ago
From where I'm sitting, the root cause is that realtime reindexing is enforced in some areas, but not others.
Meaning, it's possible to have a "burst" of activity that triggers contention on the index process.
Even if I was to batch and defer the individual updates (that we control via automated updates) that were triggering the indexer, it's still highly likely that an index table collision would occur based on "other" activity.
Eg; if the indexer is running based on schedule, and replenishment happens manually through admin, or interaction with an order, indexing gets triggered.
Now we have two processes attempting to index - one of those will "lose", leading to deadlock/stale index.

Colin Macdonald  11 days ago
I have another scenario that’s actually more pressing and highlights why I was looking at the “ignore indexing preferences in some scenarios”.

Colin Macdonald  11 days ago
The forced index process can be fired multiple times at once, leading to multiple attempts at creating/inserting/dropping the replica table happening at the same time

Colin Macdonald  11 days ago
I’ll summarise the result, but it looks like this;

Colin Macdonald  11 days ago
report.CRITICAL: SQLSTATE[23000]: Integrity constraint violation: 1062 Duplicate entry '-BED-Q-56-67614' for key 'PRIMARY', query was: INSERT  INTO `inventory_stock_3_replica`....
report.CRITICAL: SQLSTATE[42S02]: Base table or view not found: 1146 Table 'm2_core.inventory_stock_3_replica' doesn't exist, query was: INSERT  INTO `inventory_stock_3_replica

Colin Macdonald  11 days ago
(I’ve checked, the inventory data is clean)

Colin Macdonald  11 days ago
it’s just that ReindexAfterSourceItemsDeletePlugin && `ReindexAfterSourceItemsSavePlugin` force the indexer run.

Igor Miniailo:slowpoke:  11 days ago
@col btw another possibly valid approach is to move all re-indexation to dedicated  process.
So, instead of making re-indexation in the scope of process handling current request. We would just send items need to be re-indexed into the message queue. There would be only one indexer process responsible for re-indexation.
Thus we would prevent a possibility of race condition

### Fixed Issues (if relevant)
Possible Deadlocks on Reindex process

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

File: InventoryCatalogAdminUi/etc/adminhtml/system.xml [hunk 1]
```diff
@@ -33,6 +33,17 @@
...
             </group>
+            <group id="indexer" translate="label" type="text" sortOrder="600" showInDefault="1"
+                   showInWebsite="1" showInStore="1">
+                <label>Inventory Indexer Settings</label>
+                <field id="async" translate="label" type="select" sortOrder="10" showInDefault="1" showInWebsite="0"
+                       showInStore="0" canRestore="1">
+                    <label>Stock/Source reindex strategy</label>
+                    <source_model>Magento\InventoryIndexer\Model\ReindexStrategyOptions</source_model>
+                    <config_path>cataloginventory/indexer/strategy</config_path>
+                    <comment>An asynchronous queue manager must be configured for async mode.</comment>
```
Comment (maghamed) [inline InventoryCatalogAdminUi/etc/adminhtml/system.xml:44]: all labels should be approved by PO/Tech Writers

File: InventoryIndexer/Indexer/SourceItem/SourceItemReindexStrategy.php [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+declare(strict_types=1);
+
+namespace Magento\InventoryIndexer\Indexer\SourceItem;
+
+use Magento\Framework\ObjectManagerInterface;
+use Magento\InventoryIndexer\Model\IndexerConfig;
+use Magento\Framework\Exception\LocalizedException;
+
+/**
+ * Receiver of currently active reindex strategy for source items
```
Comment (maghamed) [inline InventoryIndexer/Indexer/SourceItem/SourceItemReindexStrategy.php]: should be plural - strategies

File: InventoryIndexer/Indexer/SourceItem/SourceItemReindexStrategy.php [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+    public function executeFull() : void
+    {
+        $this->getStrategy()->executeFull();
+    }
+
+    /**
+     * @param int $sourceItemId
+     * @return void
+     */
+    public function executeRow(int $sourceItemId) : void
```
Comment (maghamed) [inline InventoryIndexer/Indexer/SourceItem/SourceItemReindexStrategy.php]: do our static tests allow this extra space after closing parenthesis `) : void`

File: InventoryIndexer/Indexer/SourceItem/SourceItemReindexStrategy.php [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+    {
+        $this->getStrategy()->executeList([$sourceItemId]);
+    }
+
+    /**
+     * @return mixed
```
Comment (maghamed) [inline InventoryIndexer/Indexer/SourceItem/SourceItemReindexStrategy.php]: like it!

File: InventoryIndexer/Indexer/SourceItem/Strategy/Async.php [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+
+use Magento\Framework\MessageQueue\PublisherInterface;
+use Magento\InventoryIndexer\Indexer\Stock\StockIndexer;
+
+/**
+ * Reindex source items asynchronously.
+ */
+class Async
+{
+    private const TOPIC_SOURCE_ITEMS_INDEX = "inventory.indexer.sourceItem";
```
Comment (maghamed) [inline InventoryIndexer/Indexer/SourceItem/Strategy/Async.php:21]: the description should be added.
single quotes should be used instead

File: InventoryIndexer/Indexer/Stock/Strategy/Async.php [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+
+use Magento\Framework\MessageQueue\PublisherInterface;
+use Magento\InventoryIndexer\Indexer\Stock\GetAllStockIds;
+
+/**
+ * Reindex stocks asynchronously.
+ */
+class Async
+{
+    private const TOPIC_STOCK_INDEX = "inventory.indexer.stock";
```
Comment (maghamed) [inline InventoryIndexer/Indexer/Stock/Strategy/Async.php:20]: description and single quotes

File: InventoryIndexer/Model/IndexerConfig.php [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+namespace Magento\InventoryIndexer\Model;
+
+use Magento\Framework\App\Config\ScopeConfigInterface;
+
+/**
+ * Reindex strategy config provider.
+ */
+class IndexerConfig
+{
+    private const XML_PATH_INDEXER_STRATEGY = 'cataloginventory/indexer/strategy';
```
Comment (maghamed) [inline InventoryIndexer/Model/IndexerConfig.php:21]: description required

File: InventoryIndexer/Model/IndexerConfig.php [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+declare(strict_types=1);
+
+namespace Magento\InventoryIndexer\Model;
+
+use Magento\Framework\App\Config\ScopeConfigInterface;
+
+/**
+ * Reindex strategy config provider.
+ */
+class IndexerConfig
```
Comment (maghamed) [inline InventoryIndexer/Model/IndexerConfig.php:16]: probably this class should be declared as `@api`

File: InventoryIndexer/Model/ReindexStrategyOptions.php [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+{
+    /**
+     * @inheritdoc
+     */
+    public function toOptionArray()
+    {
+        return [
+            [
+                'value' => 'sync',
+                'label' => __('Synchronous'),
```
Comment (maghamed) [inline InventoryIndexer/Model/ReindexStrategyOptions.php:22]: all labels should be approved by PO/Dev Docs

File: InventoryIndexer/Indexer/SourceItem/SourceItemReindexStrategy.php (hunk 3 above)
Comment (sidolov) [reply InventoryIndexer/Indexer/SourceItem/SourceItemReindexStrategy.php]: Yes, it's acceptable

**Comment by sidolov** [comment]
@magento run all tests

**Comment by sidolov** [comment]
@magento run Functional Tests B2B

**Comment by sidolov** [comment]
Approval ticket: https://jira.corp.magento.com/browse/MC-34190

**Comment by sidolov** [comment]
@magento run Functional Tests B2B

**Pull request merged**